sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -1037,6 +1037,9 @@ object SimplifyCasts extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan.transformAllExpressionsWithPruning(
     _.containsPattern(CAST), ruleId) {
     case Cast(e, dataType, _, _) if e.dataType == dataType => e
+    case c1 @ Cast(Cast(e, d2: DecimalType, _, _), d1: DecimalType, _, _)
+        if d1.isWiderThan(d2) && d2.isWiderThan(e.dataType) =>
```

Comment:
can we generalize it? I think the general idea is we can combine to "upcast" into one.

Reply:
OK